codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java

```diff
@@ -252,7 +252,7 @@ public int onDataRead(final ChannelHandlerContext ctx, int streamId, ByteBuf dat
             Http2LocalFlowController flowController = flowController();
             try {
                 // If we should apply flow control, do so now.
-                if (shouldApplyFlowControl) {
+                if (shouldApplyFlowControl && bytesToReturn > 0) {
```

Comment:
Agreed ... I think we can safely remove that from the if statement.  We should always be applying flow control if `shouldApplyFlowControl` is true.